workbench-core/datasets/src/dataSetService.ts

```diff
@@ -718,52 +736,81 @@ export class DataSetService {
       );
     }
 
-    const readOnly = permissionsData.permissions.some(({ accessLevel }) => accessLevel === 'read-only');
-
-    if (_.find(targetDS.externalEndpoints, (ep) => ep === externalEndpointName)) {
-      throw new EndPointExistsError(`'${externalEndpointName}' already exists in '${dataSetId}'.`);
-    }
+    const accessLevel = this._getMinimumAccessLevel(permissionsData.permissions);
 
     const { data: connectionsData } = await storageProvider.addExternalEndpoint({
       name: targetDS.storageName,
       path: targetDS.path,
       externalEndpointName,
       ownerAccountId: targetDS.awsAccountId!,
-      accessLevel: readOnly ? 'read-only' : 'read-write',
+      accessLevel,
       externalRoleName,
       kmsKeyArn,
       vpcId
     });
 
-    const endPointParam: ExternalEndpoint = {
+    const endpointParam: CreateExternalEndpoint = {
       name: externalEndpointName,
-      dataSetId: targetDS.id!,
+      dataSetId: targetDS.id,
       dataSetName: targetDS.name,
       path: targetDS.path,
       endPointUrl: connectionsData.connections.endPointUrl,
-      endPointAlias: connectionsData.connections.endPointAlias
+      endPointAlias: connectionsData.connections.endPointAlias,
+      accessLevel
     };
 
     if (externalRoleName) {
-      endPointParam.allowedRoles = [externalRoleName];
+      endpointParam.allowedRoles = [externalRoleName];
     }
 
-    const endPoint: ExternalEndpoint = await this._dbProvider.addExternalEndpoint(endPointParam);
+    const endpoint: ExternalEndpoint = await this._dbProvider.addExternalEndpoint(endpointParam);
 
     if (!targetDS.externalEndpoints) {
       targetDS.externalEndpoints = [];
     }
 
-    targetDS.externalEndpoints.push(endPoint.id!);
+    targetDS.externalEndpoints.push(endpoint.id);
 
     await this._dbProvider.updateDataSet(targetDS);
 
-    const mountObject = this._generateMountObject(
-      endPoint.dataSetName,
-      endPoint.endPointAlias!,
-      endPoint.path,
-      endPoint.id!
+    return {
+      data: {
+        mountObject: this._generateMountObject(
+          endpoint.dataSetName,
+          endpoint.endPointAlias,
+          endpoint.path,
+          endpoint.id
+        )
+      }
+    };
+  }
+
+  private _getMinimumAccessLevel(permissions: DataSetPermission[]): DataSetsAccessLevel {
+    if (permissions.some(({ accessLevel }) => accessLevel === 'read-only')) {
+      return 'read-only';
+    }
+    if (permissions.some(({ accessLevel }) => accessLevel === 'read-write')) {
+      return 'read-write';
+    }
+    throw new NotAuthorizedError('No permissions exist');
+  }
+
+  private async _getAllUserPermissionsForDataset(
```

Comment:
?

```
  public async getAllDataSetAccessPermissions(
    datasetId: string,
    pageToken?: string
  ): Promise<PermissionsResponse> 
```
(`wbcDataSetsAuthorizationPlugin.ts`)

Reply:
Might be a misleading function name, this takes an authorized user and gets all the permissions it has on a single dataset (`USER` and `GROUP` permissions)

Reply:
renamed to make it clearer